src/lib/actions/AbstractRunAction.ts

```diff
@@ -0,0 +1,96 @@
+
+	protected constructor(display: Display, inputsResolver: InputsResolver, runOptionsFactory: RunOptionsFactory,
+						  engineOptionsFactory: EngineOptionsFactory) {
+		this.display = display;
+		this.inputsResolver = inputsResolver;
+		this.runOptionsFactory = runOptionsFactory;
+		this.engineOptionsFactory = engineOptionsFactory;
+	}
+
+	async validateInputs(inputs: Inputs): Promise<void> {
+		const fh = new FileHandler();
+		// If there's a --projectdir flag, its entries must be non-glob paths pointing to existing directories.
+		if (inputs.projectdir) {
+			// TODO: MOVE AWAY FROM ALLOWING AN ARRAY OF DIRECTORIES HERE AND ERROR IF THERE IS MORE THAN ONE DIRECTORY
+			for (const dir of (inputs.projectdir as string[])) {
+				if (globby.hasMagic(dir)) {
+					throw new SfError(getMessage(Bundle.CommonRun, 'validations.projectdirCannotBeGlob'));
+				} else if (!(await fh.exists(dir))) {
+					throw new SfError(getMessage(Bundle.CommonRun, 'validations.projectdirMustExist'));
+				} else if (!(await fh.stats(dir)).isDirectory()) {
+					throw new SfError(getMessage(Bundle.CommonRun, 'validations.projectdirMustBeDir'));
+				}
+			}
+		}
+		// If the user explicitly specified both a format and an outfile, we need to do a bit of validation there.
+		if (inputs.format && inputs.outfile) {
+			const inferredOutfileFormat: OUTPUT_FORMAT = inferFormatFromOutfile(inputs.outfile);
+			// For the purposes of this validation, we treat junit as xml.
+			const chosenFormat: string = inputs.format === 'junit' ? 'xml' : inputs.format as string;
+			// If the chosen format is TABLE, we immediately need to exit. There's no way to sensibly write the output
+			// of TABLE to a file.
+			if (chosenFormat === OUTPUT_FORMAT.TABLE) {
+				throw new SfError(getMessage(Bundle.CommonRun, 'validations.cannotWriteTableToFile', []));
+			}
+			// Otherwise, we want to be liberal with the user. If the chosen format doesn't match the outfile's extension,
+			// just log a message saying so.
+			if (chosenFormat !== inferredOutfileFormat) {
+				this.display.displayInfo(getMessage(Bundle.CommonRun, 'validations.outfileFormatMismatch', [inputs.format as string, inferredOutfileFormat]));
+			}
+		}
+	}
+
+	async run(inputs: Inputs): Promise<AnyJson> {
+		const filters: RuleFilter[] = new RuleFilterFactoryImpl().createRuleFilters(inputs);
+		const targetPaths: string[] = this.inputsResolver.resolveTargetPaths(inputs);
+		const runOptions: RunOptions = this.runOptionsFactory.createRunOptions(inputs);
+		const engineOptions: EngineOptions = this.engineOptionsFactory.createEngineOptions(inputs);
+
+		// TODO: Inject RuleManager as a dependency to improve testability by removing coupling to runtime implementation
+		const ruleManager: RuleManager = await Controller.createRuleManager();
+
+		let output: RecombinedRuleResults = null;
+		try {
+			output = await ruleManager.runRulesMatchingCriteria(filters, targetPaths, runOptions, engineOptions);
```

Comment:
I can't remember who said this quote, but it always sticks with me:
"Software Design is the art of drawing boundaries".

For the run and run dfa commands, this call to runRulesMatchingCriteria seemed like a major boundary where we have finished processing the input flags and begin more heavy run specific business logic.

You'll see in this refactoring work that I came up the services:
RuleFilterFactory, InputsResolver, RunOptionsFactory, EngineOptionsFactory. 

Each of these are interfaces so that we can control the input dependencies of this boundary. I'm not sure why these are separate (I haven't studied the code enough yet to learn), but this then makes us have the following flow:

user --> [CLI LAYER] --> inputs and display

inputs and disiplay --> [INPUT PROCESSING LAYER] --> rule filters, target paths, run options, engine options

 rule filters, target paths, run options, engine options --> [RUN LAYER (May be doing more than it needs to)] --> results
 
 results --> [OUTPUT LAYER] --> output